### PR TITLE
Implement support for non-flat directory structure

### PR DESCRIPTION
**Pull request by MunGell**
Hi there,

Would love to hear your feedback / code review as this is my first public elisp code.

The implementation already supports nested directories for drafts/posts.
However, I am still thinking of the best approach to develop @neeasade 's idea of having a separate function to generate URLs.
At the moment user can re-define `org-static-blog-get-url` to achieve that, but I believe it would be better to split this function into two: one that does necessary changes to the file path (e.g. changing `.org` to `.html`) and another to generate the rest of the URL path. The second one should be overridable.

Related to issue #10

### CODE REVIEW

**Comment by MunGell** [comment]
Added separate function to generate post URLs. 

Example of a custom URL generation function:

```elisp
(defun org-static-blog-post-url (post-filename post-datetime)
  (concat (format-time-string "%Y/%m/%d" post-datetime)
          "/"
          (file-name-nondirectory post-filename)))
```

In this example, for a post `~/blog/posts/post1.org` dated with today's date it will generate `/2019/08/27/post1.html`

There is a potential problem with relative links to static assets and auto-generated links like tags – if user uses `/` in their generated URLs, just like I did in the example above, the links get broken.

To solve this for both root- and sub-directory-hosted blogs we will need to update links in all autogenerated HTML snippets to concat `org-static-blog-publish-url`.

Suggestions welcome

**Comment by MunGell** [comment]
Added absolute path to all auto-generated links, so all links to tags etc. are functional now.

**Comment by MunGell** [comment]
Please note, that `directory-files-recursively` was implemented in Emacs 25, so minimal Emacs version will need to be bumped

**Comment by bastibe** [comment]
Thank you for this pull request. And sorry for the late response, I was on vacation.

This seems really cool!

Could you summarize your changes, so I can get an overview of what you have built? Does your pull request break existing functionality?

**Comment by MunGell** [comment]
Hi @bastibe 

Sure, I saw your status, so didn't expect a quick response.

I was trying to make my commit messages descriptive, here is a list of changes I've made:

#### `org-static-blog-get-post-filenames` and `org-static-blog-get-draft-filenames`

These functions were changed to use `directory-files-recursively` instead of `directory-files` to search sub-directories for `.org$` files.

Function `directory-files-recursively` was implemented in Emacs 25, so version requirement on line 6 will need to be updated (can be added to this PR if needed).

#### `org-static-blog-file-buffer`

Function was updated to create sub-directories in public directory as per post URL. Supports recursive directory creation.

#### `org-static-blog-get-url`

This function is now changed to pass cleaned post URL (from `org-static-blog-prepare-post-url`) and post datetime (from `org-static-blog-get-date`) to URL generation function that could be overridden by the user (`org-static-blog-post-url`).

It also concatenates `.html` to the end of each URL.

#### `org-static-blog-prepare-post-url`

New function to clean the URL:
* Removes `.org` suffix in the URL (later substituted with `.html` in `org-static-blog-get-url`)
* Removes parent directories (absolute paths to posts and drafts directories) from the URL

#### `org-static-blog-post-url`

New function to generate custom post URLs, can be overridden by the user.
By default returns the inputted cleaned post URL to ensure backwards compatibility – for root-level files will return filename.

#### `org-static-blog-matching-publish-filename`

This function is modified to use updated `org-static-blog-get-url` instead of assuming flat directory structure with org files in `/posts/` root.

---

The rest of the changes add `org-static-blog-publish-url` to the URLs in HTML to avoid problems with relative paths, e.g. if users decide they want to have URLs like `/YYYY/mm/dd/post-name.html` relative path to `archive.html` would resolve to `/YYYY/mm/dd/archive.html`.

---

Example of custom URL generation can be this:

```elisp
(defun org-static-blog-post-url (post-filename post-datetime)
  (concat (format-time-string "%Y/%m/%d" post-datetime)
          "/"
          (file-name-nondirectory post-filename)))
```

which generates URLs like `/YYYY/mm/dd/post-name.html`.

---

### Downsides

There is a couple of things I would like to change in future versions of this functionality:
* I find function names a little bit confusing (e.g. `org-static-blog-get-url` vs `org-static-blog-post-url`). Didn't change them in this PR to prevent confusion and make changes more visible. Can be subject to the next PR.
* I am not a big fan of how `.org$` replacement with `.html$` functionality is spread between 2 functions. It also prevents advanced customization of URLs by assuming all URLs end with `.html`. I am still thinking of a better approach there with will create a PR if will come up with something elegant.

**Comment by bastibe** [comment]
Thank you very much. I like it!

However, I agree with you that the function names are now confusing, and I think we should rename them before merging. Feel free to mess with existing names as well, if that improves things.

Also, shouldn't `org-static-blog-post-url` be a `defcustom`?

**Comment by MunGell** [comment]
Thanks @bastibe 

I believe `org-static-blog-post-url` needs to stay as a function – it should be able to accommodate any URL generation logic defined by the user. Although, at the moment it returns just filepath back, it is possible to override it with something like that (example from the above comment):

```elisp
(defun org-static-blog-post-url (post-filename post-datetime)
  (concat (format-time-string "%Y/%m/%d" post-datetime)
          "/"
          (file-name-nondirectory post-filename)))
```

I will review naming of functions and variables and will update the PR

File: org-static-blog.el [hunk 1]
```diff
@@ -302,27 +302,36 @@ Preamble and Postamble are excluded, too."
...
-            (file-truename org-static-blog-posts-directory)
-            "\\|"
-            (file-truename org-static-blog-drafts-directory)
-            "\\)")
-    ""
-    post-filename)))
-
-(defun org-static-blog-post-url (post-filename post-datetime)
-  "Generate arbitrary post URL. This function is overridable by user"
+(defun org-static-blog-get-full-url (relative-url)
```
Comment (bastibe) [inline org-static-blog.el]: I think this should be called `org-static-blog-get-absolute-url`.

File: org-static-blog.el [hunk 2]
```diff
@@ -302,27 +302,36 @@ Preamble and Postamble are excluded, too."
...
+  "Return absolute URL based on the RELATIVE-URL passed to the function."
+  (concat org-static-blog-publish-url relative-url))
+
+(defun org-static-blog-get-post-url (post-filename)
+  "Generate full URL to the published POST-FILENAME."
+  (org-static-blog-get-full-url
+          (org-static-blog-get-post-public-path post-filename)))
+
+(defun org-static-blog-get-post-public-path (post-filename)
+  "Return post filepath in public directory."
```
Comment (bastibe) [inline org-static-blog.el]: How about we add a short example to the docstrings of each of these functions?

File: org-static-blog.el [hunk 3]
```diff
@@ -302,27 +302,36 @@ Preamble and Postamble are excluded, too."
...
+                             (concat "^\\("
+                                     (file-truename org-static-blog-posts-directory)
+                                     "\\|"
+                                     (file-truename org-static-blog-drafts-directory)
+                                     "\\)")
+                             ""
+                             post-filename)))
+
+(defun org-static-blog-generate-post-path (post-filename post-datetime)
+  "Generate post public path (URL) based on POST-FILENAME and PORT-DATETIME.
```
Comment (bastibe) [inline org-static-blog.el]: As far as I can tell, the other functions either end on `-path` or `-url`, depending on whether they return a local file path or a public url. This one seems to be the odd one out.

File: org-static-blog.el [hunk 4]
```diff
@@ -302,10 +302,81 @@ Preamble and Postamble are excluded, too."
...
+For example, when `org-static-blog-publish-url` is set to 'https://example.com/'
+and `relative-url` is passed as 'archive.html' then the function
+will return 'https://example.com/archive.html'."
+  (concat org-static-blog-publish-url relative-url))
+
+(defun org-static-blog-get-post-url (post-filename)
+  "Returns absolute URL to the published POST-FILENAME.
+
+This function concatenates publish URL and generated custom filepath to the
+published HTML version of the post."
```
Comment (bastibe) [inline org-static-blog.el]: Could you add an explicit example here, like in `org-static-blog-get-absolute-url`?

File: org-static-blog.el [hunk 5]
```diff
@@ -302,10 +302,81 @@ Preamble and Postamble are excluded, too."
...
+  (org-static-blog-get-absolute-url
+          (org-static-blog-get-post-public-path post-filename)))
+
+(defun org-static-blog-get-post-public-path (post-filename)
+  "Returns post filepath in public directory.
+
+This function retrieves relative path to the post file in posts or drafts
+directories, the date of the post from its contents and then passes it to
+`org-static-blog-generate-post-path` to generate custom filepath for the published
+HTML version of the post."
```
Comment (bastibe) [inline org-static-blog.el]: Could you add an explicit example here, like in `org-static-blog-get-absolute-url`?

**Comment by MunGell** [comment]
Hi @bastibe 

Unfortunately, I will not be able to spend much time on this PR anymore. Please feel free to take over from here if any further changes are required (I believe you have edit access to my branch).

Thanks!

**Comment by bastibe** [comment]
No problem. Thank you for your work so far!

**Comment by winny-** [comment]
Hello! After bisecting master for regressions, I found there are a couple issues with this PR:

- This appears to break `.dir-locals.el` (in the blog root). They are no longer applied on `org-static-blog-publish`.
- My `org-static-blog-publish-url` is set to something like `https://example.com/`, which unfortunately breaks previewing the site locally using a httpd (`python -m http.server`). Since all the URLs are prefixed with `https://example.com/` now, most links (RSS, `a` elements) are incorrect when viewing the site from anywhere other than `https://example.com/`. A quick and dirty fix on my end would be to set `org-static-blog-publish-url` to `/` but I am unsure if this'll cause further breakage when deployed to my actual `example.com` host. A better solution could be to use relative URLs, to ensure structure is traverse-able and not necessarily rooted at `org-static-blog-publish-url`

Any thoughts? This is a really cool feature. I wish I had noticed these issues earlier! :)

**Comment by MunGell** [comment]
Hi @winny- 

I think you are right, this is not an ideal behaviour (I personally just change `org-static-blog-publish-url` to `http://localhost` and don't commit this to the repo).

The original idea for absolute urls was to solve for situations when blog is hosted in a subdirectory or when complex blog post url generation is implemented (that involves slashes).

The solution could be in splitting `org-static-blog-publish-url` into 2 properties:
* `org-static-blog-publish-domain` - that defines the domain of the blog
* `org-static-blog-publish-path` - that defines path to the subdirectory of the blog. This should default to `/` for compatibility reasons.

Let me know what you think

**Comment by bastibe** [comment]
These issues seem unrelated to the current pull request. I would prefer them to be discussed as a separate issue.
